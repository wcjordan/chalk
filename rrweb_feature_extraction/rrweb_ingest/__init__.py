"""
rrweb Input Ingestion & Preprocessing Module

This module processes rrweb session recordings and transforms them into
clean, structured chunks ready for downstream feature extraction and behavior analysis.
"""

__version__ = "0.1.0"
