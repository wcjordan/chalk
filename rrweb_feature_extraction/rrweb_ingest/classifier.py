"""
Event Classification for rrweb session data.

This module provides functionality to classify rrweb events into different categories
based on their type field. Events are separated into snapshots (full DOM captures),
interactions (incremental changes), and other event types for downstream processing.
"""

from typing import List, Tuple


def classify_events(events: List[dict]) -> Tuple[List[dict], List[dict], List[dict]]:
    """
    Classify rrweb events into snapshots, interactions, and others based on event type.

    Takes a sorted list of rrweb event dictionaries and separates them into three
    categories based on their 'type' field:
    - Snapshots: Full DOM snapshots (type == 2)
    - Interactions: Incremental changes (type == 3)
    - Others: All remaining event types (Meta, Custom, Plugin, etc.)

    Args:
        events: List of rrweb event dictionaries, typically sorted by timestamp

    Returns:
        Tuple containing three lists in order: (snapshots, interactions, others)

    Raises:
        KeyError: If any event is missing the required 'type' field
    """
    snapshots = []
    interactions = []
    others = []

    for event in events:
        # Access 'type' field - will raise KeyError if missing
        event_type = event["type"]

        if event_type == 2:
            # FullSnapshot events - complete DOM captures
            snapshots.append(event)
        elif event_type == 3:
            # IncrementalSnapshot events - DOM mutations, mouse moves, inputs, etc.
            interactions.append(event)
        else:
            # All other event types (Meta, Custom, Plugin, etc.)
            others.append(event)

    return snapshots, interactions, others
