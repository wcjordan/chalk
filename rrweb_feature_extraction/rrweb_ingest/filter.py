"""
Noise-Filtering Framework for rrweb session data.

This module provides functionality to identify and remove low-signal events from
rrweb chunks, including mousemove noise, micro-scrolls, trivial DOM mutations,
and duplicate events. This helps focus downstream processing on meaningful
user interactions.
"""

from typing import List, Set, Tuple


def is_low_signal(event: dict, micro_scroll_threshold: int = 20) -> bool:
    """
    Returns True if the event should be dropped as low-signal noise.

    Identifies events that carry little meaningful information about user behavior,
    such as cursor-only movements, tiny scrolls, and trivial DOM changes.

    Args:
        event: rrweb event dictionary with 'type', 'timestamp', and 'data' fields
        micro_scroll_threshold: Minimum scroll distance in pixels to be considered
                               meaningful (default: 20px)

    Returns:
        True if the event should be filtered out as noise, False if it should be kept

    Examples:
        >>> # Mousemove event (noise)
        >>> event = {"type": 3, "data": {"source": 1}, "timestamp": 1000}
        >>> is_low_signal(event)
        True

        >>> # Significant scroll (keep)
        >>> event = {"type": 3, "data": {"source": 3, "y": 50}, "timestamp": 1000}
        >>> is_low_signal(event)
        False

        >>> # Micro-scroll (noise)
        >>> event = {"type": 3, "data": {"source": 3, "y": 10}, "timestamp": 1000}
        >>> is_low_signal(event)
        True
    """
    # Only filter IncrementalSnapshot events (type == 3)
    if event.get("type") != 3:
        return False

    data = event.get("data", {})
    source = data.get("source")

    # Drop mousemove events (source == 1)
    if source == 1:
        return True

    # Drop micro-scrolls (source == 3 with small delta)
    if source == 3:
        # Check for scroll distance in x or y direction
        x_delta = abs(data.get("x", 0))
        y_delta = abs(data.get("y", 0))

        # If both deltas are below threshold, consider it a micro-scroll
        if x_delta < micro_scroll_threshold and y_delta < micro_scroll_threshold:
            return True

    # Drop trivial DOM mutations (source == 0)
    if source == 0:
        # Check for trivial mutations - this is a simplified heuristic
        # In practice, this would need more sophisticated analysis
        adds = data.get("adds", [])
        removes = data.get("removes", [])
        texts = data.get("texts", [])
        attributes = data.get("attributes", [])

        # If no significant changes, consider it trivial
        if not adds and not removes and not texts and not attributes:
            return True

        # If only minor attribute changes (like style updates), consider trivial
        if not adds and not removes and not texts and len(attributes) == 1:
            attr_change = attributes[0]
            # Simple heuristic: single style attribute changes are often trivial
            if attr_change.get("attributes", {}).get("style"):
                return True

    return False


def clean_chunk(events: List[dict], micro_scroll_threshold: int = 20) -> List[dict]:
    """
    Removes low-signal and duplicate events from a chunk.

    Applies noise filtering to remove events that don't contribute meaningful
    information about user behavior, and deduplicates identical events to
    reduce redundancy.

    Args:
        events: List of rrweb event dictionaries to clean
        micro_scroll_threshold: Minimum scroll distance in pixels to be
                                considered meaningful (default: 20px)

    Returns:
        List of cleaned events with noise and duplicates removed, preserving
        the original order of remaining events

    Examples:
        >>> events = [
        ...     {"type": 3, "data": {"source": 1}, "timestamp": 1000},  # mousemove (noise)
        ...     {"type": 3, "data": {"source": 2, "id": 5}, "timestamp": 2000},  # click
        ...     {"type": 3, "data": {"source": 2, "id": 5}, "timestamp": 2000},  # duplicate
        ... ]
        >>> clean_chunk(events)
        [{"type": 3, "data": {"source": 2, "id": 5}, "timestamp": 2000}]
    """
    if not events:
        return []

    cleaned = []
    seen_signatures: Set[Tuple] = set()

    for event in events:
        # Skip low-signal events
        if is_low_signal(event, micro_scroll_threshold):
            continue

        # Create a signature for deduplication
        # Use type, source, timestamp, and target id if available
        event_type = event.get("type")
        data = event.get("data", {})
        source = data.get("source")
        timestamp = event.get("timestamp")
        target_id = data.get("id")  # Target element ID for interactions

        signature = (event_type, source, timestamp, target_id)

        # Skip if we've seen this exact event before
        if signature in seen_signatures:
            continue

        seen_signatures.add(signature)
        cleaned.append(event)

    return cleaned
