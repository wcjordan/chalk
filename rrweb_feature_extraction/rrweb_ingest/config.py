"""
Configuration settings for rrweb ingestion pipeline.

This module defines default values for all configurable parameters used throughout
the ingestion pipeline. These defaults can be overridden by passing parameters
to individual functions or by modifying the values in this module.

Configuration Categories:
- Chunking thresholds: Control when new chunks are created
- Noise filtering: Control what events are considered low-signal
- Extensibility: Support for custom filtering logic

Usage Examples:
    # Use default values
    from rrweb_ingest.config import MAX_GAP_MS
    
    # Override in function calls
    chunks = segment_into_chunks(interactions, snapshots, max_gap_ms=5000)
    
    # Custom noise filter
    def custom_filter(event):
        return event.get("data", {}).get("source") == 99
    
    cleaned = clean_chunk(events, custom_filters=[custom_filter])
"""

# Chunking Configuration
# ======================

# Maximum time gap in milliseconds between consecutive interactions
# before starting a new chunk. Larger gaps typically indicate separate
# user workflows or sessions.
MAX_GAP_MS = 10_000

# Maximum number of events per chunk before forcing a split.
# Prevents chunks from becoming too large for efficient processing.
MAX_EVENTS = 1000

# Maximum duration in milliseconds for a single chunk.
# Ensures chunks represent reasonable time windows for analysis.
MAX_DURATION_MS = 30_000

# Noise Filtering Configuration
# =============================

# Minimum scroll distance in pixels to be considered meaningful.
# Scrolls below this threshold are typically considered noise
# (accidental micro-scrolls, scroll wheel sensitivity, etc.).
MICRO_SCROLL_THRESHOLD = 20

# DOM Mutation Filtering
# ======================

# Whether to filter empty DOM mutations (no adds, removes, texts, or attributes)
FILTER_EMPTY_MUTATIONS = True

# Whether to filter single style-only attribute changes as trivial
FILTER_STYLE_ONLY_MUTATIONS = True

# Input Event Filtering
# ====================

# Whether to filter input events that don't have corresponding blur/submit
# within the same chunk (incomplete inputs are often noise)
FILTER_INCOMPLETE_INPUTS = False  # Not yet implemented

# Extensibility Configuration
# ===========================

# Default custom filter functions (empty by default)
# Users can add custom predicates here or pass them to functions
DEFAULT_CUSTOM_FILTERS = []

# Configuration Validation
# ========================

def validate_config():
    """
    Validate that all configuration values are within reasonable ranges.
    
    Raises:
        ValueError: If any configuration value is invalid
    """
    if MAX_GAP_MS <= 0:
        raise ValueError("MAX_GAP_MS must be positive")
    
    if MAX_EVENTS <= 0:
        raise ValueError("MAX_EVENTS must be positive")
    
    if MAX_DURATION_MS <= 0:
        raise ValueError("MAX_DURATION_MS must be positive")
    
    if MICRO_SCROLL_THRESHOLD < 0:
        raise ValueError("MICRO_SCROLL_THRESHOLD must be non-negative")


# Validate configuration on import
validate_config()
