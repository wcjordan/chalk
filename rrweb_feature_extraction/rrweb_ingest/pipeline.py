"""
End-to-End Ingest Pipeline for rrweb session data.

This module provides the main entry point for processing rrweb session recordings.
It orchestrates the complete pipeline from raw JSON loading through chunk normalization,
integrating all preprocessing components into a single cohesive workflow.

The pipeline performs the following steps:
1. Load and validate raw rrweb JSON data
2. Classify events into snapshots, interactions, and others
3. Segment interactions into logical chunks based on boundaries and limits
4. Filter out noise and duplicate events from each chunk
5. Normalize chunks into standardized objects with metadata
"""
import os
from collections import defaultdict
import sys
from typing import List

from rrweb_ingest.loader import load_events
from rrweb_ingest.classifier import classify_events
from rrweb_ingest.segmenter import segment_into_chunks
from rrweb_ingest.filter import clean_chunk
from rrweb_ingest.normalizer import normalize_chunk
from rrweb_ingest.models import Chunk


def ingest_session(
    session_id: str,
    filepath: str,
) -> List[Chunk]:
    """
    Load, classify, segment, filter, and normalize an rrweb session into Chunks.

    This is the main entry point for processing rrweb session recordings. It takes
    a raw JSON file and transforms it through the complete preprocessing pipeline
    to produce a list of normalized, cleaned chunks ready for feature extraction.

    The pipeline integrates all preprocessing components:
    1. Loads and validates the JSON session file
    2. Classifies events by type (snapshots vs interactions vs others)
    3. Segments interactions into chunks based on multiple criteria
    4. Filters noise and duplicates from each chunk
    5. Normalizes chunks with metadata and standardized identifiers

    Args:
        session_id: Unique identifier for this session, used in chunk IDs
        filepath: Path to the rrweb JSON session file to process

    Returns:
        List of normalized Chunk objects, each containing cleaned events and
        metadata. Chunks are ordered by their temporal sequence in the session.

    Raises:
        FileNotFoundError: If the specified filepath does not exist
        JSONDecodeError: If the file contains invalid JSON syntax
        ValueError: If the JSON structure is invalid or session_id is empty
        KeyError: If events are missing required fields

    Examples:
        >>> chunks = ingest_session(
        ...     "user_session_123",
        ...     "/path/to/session.json",
        ... )
        >>> len(chunks)
        3
        >>> chunks[0].chunk_id
        'user_session_123-chunk000'
        >>> chunks[0].metadata['num_events']
        45
    """
    # Validate session_id
    if not session_id:
        raise ValueError("session_id cannot be empty")

    # Step 1: Load and validate events from JSON file
    events = load_events(filepath)

    # Step 2: Classify events into categories
    snapshots, interactions, _ = classify_events(events)

    # Step 3: Segment interactions into raw chunks
    raw_chunks = segment_into_chunks(interactions, snapshots)

    # Step 4 & 5: Clean and normalize each chunk
    normalized_chunks = []
    for chunk_index, raw_chunk in enumerate(raw_chunks):
        # Filter noise and duplicates
        cleaned_events = clean_chunk(raw_chunk)

        # Skip empty chunks after cleaning
        if not cleaned_events:
            continue

        # Normalize into Chunk object
        chunk = normalize_chunk(cleaned_events, session_id, chunk_index)
        normalized_chunks.append(chunk)

    return normalized_chunks


if __name__ == "__main__":
    session_path = "../session_stitching/output_sessions"
    chunk_sizes = defaultdict(int)
    for filename in os.listdir(session_path):
        if filename.endswith(".json"):
            session_id = filename.split(".")[0]
            filepath = os.path.join(session_path, filename)
            try:
                chunks = ingest_session(session_id, filepath)
                chunk_sizes[len(chunks)] += 1
            except Exception as e:
                print(f"Error processing {filename}: {e}")

    print("Chunk size distribution:")
    for size, count in sorted(chunk_sizes.items()):
        print(f"{size} events: {count} sessions")
    print(f"Total sessions processed: {sum(chunk_sizes.values())}")
