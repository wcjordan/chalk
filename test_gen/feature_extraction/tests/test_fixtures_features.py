"""
Fixtures-based integration tests for the feature extraction pipeline.

Tests the complete pipeline using sample session data to verify end-to-end
functionality and accurate feature detection across all major event types.
"""

import json
import os
import pytest
from rrweb_ingest.pipeline import ingest_session
from feature_extraction.dom_state import init_dom_state
from feature_extraction.pipeline import extract_features


@pytest.fixture(name="sample_session_path")
def fixture_sample_session_path():
    """Fixture providing the path to the sample session JSON file."""
    current_dir = os.path.dirname(__file__)
    return os.path.join(current_dir, "fixtures", "sample_session.json")


@pytest.fixture(name="sample_featurechunk_path")
def fixture_sample_featurechunk_path():
    """Fixture providing the path to the expected feature chunk JSON file."""
    current_dir = os.path.dirname(__file__)
    return os.path.join(current_dir, "fixtures", "sample_featurechunk.json")


@pytest.fixture(name="sample_session_data")
def fixture_sample_session_data(sample_session_path):
    """Fixture providing the loaded sample session data."""
    with open(sample_session_path, "r", encoding="utf-8") as f:
        return json.load(f)


@pytest.fixture(name="expected_featurechunk")
def fixture_expected_featurechunk(sample_featurechunk_path):
    """Fixture providing the expected feature chunk structure."""
    with open(sample_featurechunk_path, "r", encoding="utf-8") as f:
        return json.load(f)


def test_sample_session_loads_successfully(sample_session_data):
    """Test that the sample session JSON loads and has expected structure."""
    assert isinstance(sample_session_data, list)
    assert len(sample_session_data) >= 15  # Should have ~15-25 events
    assert len(sample_session_data) <= 25

    # Should have at least one FullSnapshot event
    full_snapshots = [event for event in sample_session_data if event.get("type") == 2]
    assert len(full_snapshots) >= 1

    # Should have various IncrementalSnapshot events
    incremental_snapshots = [
        event for event in sample_session_data if event.get("type") == 3
    ]
    assert len(incremental_snapshots) >= 10


def test_sample_session_covers_all_event_types(sample_session_data):
    """Test that the sample session covers all major event types."""
    # Collect all source types from IncrementalSnapshot events
    sources = set()
    for event in sample_session_data:
        if event.get("type") == 3:
            source = event.get("data", {}).get("source")
            if source is not None:
                sources.add(source)

    # Should include mousemove (1), click (2), scroll (3), mutation (0), input (5)
    expected_sources = {0, 1, 2, 3, 5}
    assert expected_sources.issubset(
        sources
    ), f"Missing sources: {expected_sources - sources}"


def test_end_to_end_feature_extraction(sample_session_path):
    """Test complete pipeline from sample session to feature extraction."""
    # Load session via ingest_session
    chunks = ingest_session("sample-session", sample_session_path)

    assert len(chunks) >= 1, "Should produce at least one chunk"

    # Process the first chunk
    first_chunk = chunks[0]

    # Initialize DOM state from FullSnapshot if available
    dom_state = {}
    if first_chunk.metadata.get("snapshot_before"):
        dom_state = init_dom_state(first_chunk.metadata["snapshot_before"])
    else:
        # Find FullSnapshot in chunk events
        for event in first_chunk.events:
            if event.get("type") == 2:
                dom_state = init_dom_state(event)
                break

    assert len(dom_state) > 0, "Should have initialized DOM state"

    # Extract features from the chunk
    feature_chunk = extract_features(first_chunk, dom_state)

    # Verify basic structure
    assert feature_chunk.chunk_id == first_chunk.chunk_id
    assert feature_chunk.start_time == first_chunk.start_time
    assert feature_chunk.end_time == first_chunk.end_time
    assert len(feature_chunk.events) == len(first_chunk.events)

    # Verify all feature categories are present
    required_features = [
        "dom_mutations",
        "interactions",
        "inter_event_delays",
        "reaction_delays",
        "ui_nodes",
        "mouse_clusters",
        "scroll_patterns",
    ]

    for feature_type in required_features:
        assert (
            feature_type in feature_chunk.features
        ), f"Missing feature type: {feature_type}"

    # Verify feature lists are non-empty where expected
    assert len(feature_chunk.features["dom_mutations"]) > 0, "Should have DOM mutations"
    assert (
        len(feature_chunk.features["interactions"]) > 0
    ), "Should have user interactions"
    assert (
        len(feature_chunk.features["inter_event_delays"]) > 0
    ), "Should have inter-event delays"
    assert (
        len(feature_chunk.features["reaction_delays"]) > 0
    ), "Should have reaction delays"
    assert len(feature_chunk.features["ui_nodes"]) > 0, "Should have UI metadata"
    assert (
        len(feature_chunk.features["mouse_clusters"]) > 0
    ), "Should have mouse clusters"
    assert (
        len(feature_chunk.features["scroll_patterns"]) > 0
    ), "Should have scroll patterns"


def test_feature_extraction_accuracy(sample_session_path):
    """Test that feature extraction produces accurate results for known sample data."""
    chunks = ingest_session("sample-session", sample_session_path)
    first_chunk = chunks[0]

    # Initialize DOM state
    dom_state = {}
    for event in first_chunk.events:
        if event.get("type") == 2:
            dom_state = init_dom_state(event)
            break

    feature_chunk = extract_features(first_chunk, dom_state)

    # Verify specific feature counts based on sample data
    # Sample session should have:
    # - 5 DOM mutations (2 attribute, 1 text, 1 add, 1 attribute)
    # - 5 user interactions (2 clicks, 1 input, 1 scroll, 1 click)
    # - 3 mouse clusters
    # - 1 scroll pattern

    assert (
        len(feature_chunk.features["dom_mutations"]) == 5
    ), f"Expected 5 DOM mutations, got {len(feature_chunk.features['dom_mutations'])}"

    assert (
        len(feature_chunk.features["interactions"]) == 5
    ), f"Expected 5 interactions, got {len(feature_chunk.features['interactions'])}"

    assert (
        len(feature_chunk.features["mouse_clusters"]) == 3
    ), f"Expected 3 mouse clusters, got {len(feature_chunk.features['mouse_clusters'])}"

    assert (
        len(feature_chunk.features["scroll_patterns"]) == 1
    ), f"Expected 1 scroll pattern, got {len(feature_chunk.features['scroll_patterns'])}"

    # Verify interaction types
    interaction_types = [i.action for i in feature_chunk.features["interactions"]]
    assert interaction_types.count("click") == 3, "Should have 3 click interactions"
    assert interaction_types.count("input") == 1, "Should have 1 input interaction"
    assert interaction_types.count("scroll") == 1, "Should have 1 scroll interaction"

    # Verify mutation types
    mutation_types = [m.mutation_type for m in feature_chunk.features["dom_mutations"]]
    assert mutation_types.count("attribute") == 3, "Should have 3 attribute mutations"
    assert mutation_types.count("text") == 1, "Should have 1 text mutation"
    assert mutation_types.count("add") == 1, "Should have 1 add mutation"


def test_ui_metadata_resolution(sample_session_path):
    """Test that UI metadata is correctly resolved for sample session nodes."""
    chunks = ingest_session("sample-session", sample_session_path)
    first_chunk = chunks[0]

    # Initialize DOM state
    dom_state = {}
    for event in first_chunk.events:
        if event.get("type") == 2:
            dom_state = init_dom_state(event)
            break

    feature_chunk = extract_features(first_chunk, dom_state)
    ui_nodes = feature_chunk.features["ui_nodes"]

    # Should have metadata for referenced nodes
    assert len(ui_nodes) >= 4, "Should have metadata for at least 4 nodes"

    # Check specific nodes from sample data
    # Node 9: email input
    if 9 in ui_nodes:
        email_node = ui_nodes[9]
        assert email_node["tag"] == "input"
        assert email_node["aria_label"] == "Email Address"
        assert email_node["data_testid"] == "email-input"
        assert "email" in email_node["dom_path"]

    # Node 10: submit button
    if 10 in ui_nodes:
        submit_node = ui_nodes[10]
        assert submit_node["tag"] == "button"
        assert submit_node["aria_label"] == "Submit Form"
        assert submit_node["data_testid"] == "submit-btn"
        assert submit_node["text"] == "Submit"


def test_timing_accuracy(sample_session_path):
    """Test that timing delays are computed accurately."""
    chunks = ingest_session("sample-session", sample_session_path)
    first_chunk = chunks[0]

    # Initialize DOM state
    dom_state = {}
    for event in first_chunk.events:
        if event.get("type") == 2:
            dom_state = init_dom_state(event)
            break

    feature_chunk = extract_features(first_chunk, dom_state)

    # Check inter-event delays
    inter_delays = feature_chunk.features["inter_event_delays"]
    assert len(inter_delays) >= 15, "Should have delays for consecutive events"

    # Check reaction delays
    reaction_delays = feature_chunk.features["reaction_delays"]
    assert len(reaction_delays) >= 3, "Should have reaction delays"

    # Verify reaction delay timing (input->mutation should be 50ms)
    input_reaction = None
    for delay in reaction_delays:
        if delay.delta_ms == 50:  # Input to mutation delay
            input_reaction = delay
            break

    assert input_reaction is not None, "Should find input->mutation reaction delay"
    assert input_reaction.delta_ms == 50


def test_mouse_clustering_accuracy(sample_session_path):
    """Test that mouse clustering produces expected results."""
    chunks = ingest_session("sample-session", sample_session_path)
    first_chunk = chunks[0]

    # Initialize DOM state
    dom_state = {}
    for event in first_chunk.events:
        if event.get("type") == 2:
            dom_state = init_dom_state(event)
            break

    feature_chunk = extract_features(first_chunk, dom_state)
    mouse_clusters = feature_chunk.features["mouse_clusters"]

    assert len(mouse_clusters) == 3, "Should have 3 mouse clusters"

    # Check first cluster (3 points over 100ms)
    first_cluster = mouse_clusters[0]
    assert first_cluster.point_count == 3
    assert first_cluster.duration_ms == 100
    assert len(first_cluster.points) == 3

    # Verify point coordinates
    assert first_cluster.points[0]["x"] == 100
    assert first_cluster.points[0]["y"] == 150
    assert first_cluster.points[2]["x"] == 140
    assert first_cluster.points[2]["y"] == 170


def test_scroll_pattern_detection(sample_session_path):
    """Test that scroll patterns are correctly detected."""
    chunks = ingest_session("sample-session", sample_session_path)
    first_chunk = chunks[0]

    # Initialize DOM state
    dom_state = {}
    for event in first_chunk.events:
        if event.get("type") == 2:
            dom_state = init_dom_state(event)
            break

    feature_chunk = extract_features(first_chunk, dom_state)
    scroll_patterns = feature_chunk.features["scroll_patterns"]

    assert len(scroll_patterns) == 1, "Should have 1 scroll pattern"

    pattern = scroll_patterns[0]
    assert pattern.delay_ms == 200, "Scroll->mutation delay should be 200ms"
    assert pattern.scroll_event["data"]["source"] == 3, "Should be scroll event"
    assert pattern.mutation_event["data"]["source"] == 0, "Should be mutation event"


def test_feature_chunk_schema_validation(sample_session_path, expected_featurechunk):
    """Test that the extracted feature chunk matches the expected schema."""
    chunks = ingest_session("sample-session", sample_session_path)
    first_chunk = chunks[0]

    # Initialize DOM state
    dom_state = {}
    for event in first_chunk.events:
        if event.get("type") == 2:
            dom_state = init_dom_state(event)
            break

    feature_chunk = extract_features(first_chunk, dom_state)

    # Compare structure with expected output
    expected = expected_featurechunk

    # Verify feature counts match expectations
    for feature_type in expected["features"]:
        actual_count = len(feature_chunk.features[feature_type])
        expected_count = len(expected["features"][feature_type])
        assert (
            actual_count == expected_count
        ), f"Feature {feature_type}: expected {expected_count}, got {actual_count}"

    # Verify specific mutation details
    actual_mutations = feature_chunk.features["dom_mutations"]
    expected_mutations = expected["features"]["dom_mutations"]

    for i, expected_mutation in enumerate(expected_mutations):
        actual_mutation = actual_mutations[i]
        assert actual_mutation.mutation_type == expected_mutation["mutation_type"]
        assert actual_mutation.target_id == expected_mutation["target_id"]
        assert actual_mutation.timestamp == expected_mutation["timestamp"]


def test_error_handling_with_malformed_session():
    """Test that the pipeline handles malformed session data gracefully."""
    # Create a minimal malformed session
    malformed_session = [
        {"type": "invalid", "timestamp": 1000},  # Invalid type
        {"timestamp": 2000},  # Missing type
        {"type": 3, "data": {}},  # Missing source
    ]

    # Write to temporary file
    import tempfile

    with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False) as f:
        json.dump(malformed_session, f)
        temp_path = f.name

    try:
        # Should not crash, but may produce empty chunks
        chunks = ingest_session("malformed-session", temp_path)

        # If chunks are produced, feature extraction should handle them gracefully
        if chunks:
            feature_chunk = extract_features(chunks[0], {})
            assert isinstance(feature_chunk.features, dict)
            assert all(
                isinstance(feature_list, (list, dict))
                for feature_list in feature_chunk.features.values()
            )

    finally:
        # Clean up temporary file
        os.unlink(temp_path)


def test_performance_with_sample_session(sample_session_path):
    """Test that feature extraction completes in reasonable time."""
    import time

    start_time = time.time()

    chunks = ingest_session("sample-session", sample_session_path)
    first_chunk = chunks[0]

    # Initialize DOM state
    dom_state = {}
    for event in first_chunk.events:
        if event.get("type") == 2:
            dom_state = init_dom_state(event)
            break

    feature_chunk = extract_features(first_chunk, dom_state)

    end_time = time.time()
    processing_time = end_time - start_time

    # Should complete in under 1 second for sample data
    assert processing_time < 1.0, f"Processing took {processing_time:.2f}s, too slow"

    # Verify results are complete
    assert len(feature_chunk.features["dom_mutations"]) > 0
    assert len(feature_chunk.features["interactions"]) > 0
