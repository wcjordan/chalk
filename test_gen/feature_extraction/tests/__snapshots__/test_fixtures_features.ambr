# serializer version: 1
# name: test_feature_extraction
  FeatureChunk(chunk_id='sample-session-chunk000', start_time=1640995200250, end_time=1640995202100, events=[{'type': 3, 'data': {'source': 2, 'type': 2, 'id': 9, 'x': 140, 'y': 170}, 'timestamp': 1640995200250}, {'type': 3, 'data': {'source': 5, 'text': 'user@example.com', 'isChecked': False, 'id': 9}, 'timestamp': 1640995200800}, {'type': 3, 'data': {'source': 0, 'texts': [{'id': 9, 'value': 'user@example.com'}], 'attributes': [{'id': 9, 'attributes': {'value': 'user@example.com'}}], 'removes': [], 'adds': []}, 'timestamp': 1640995200850}, {'type': 3, 'data': {'source': 2, 'type': 2, 'id': 10, 'x': 210, 'y': 230}, 'timestamp': 1640995201100}, {'type': 3, 'data': {'source': 0, 'texts': [], 'attributes': [{'id': 10, 'attributes': {'class': 'btn-clicked'}}], 'removes': [], 'adds': []}, 'timestamp': 1640995201300}, {'type': 3, 'data': {'source': 3, 'id': 7, 'x': 0, 'y': 100}, 'timestamp': 1640995201500}, {'type': 3, 'data': {'source': 0, 'texts': [], 'attributes': [], 'removes': [], 'adds': [{'parentId': 7, 'nextId': None, 'node': {'type': 'Element', 'tagName': 'div', 'attributes': {'class': 'success-message', 'data-testid': 'success-msg'}, 'childNodes': [{'type': 'Text', 'textContent': 'Form submitted successfully!', 'id': 13}], 'id': 12}}]}, 'timestamp': 1640995201700}, {'type': 3, 'data': {'source': 2, 'type': 2, 'id': 12, 'x': 320, 'y': 360}, 'timestamp': 1640995202100}], features={'dom_mutations': [DomMutation(mutation_type='attribute', target_id=9, details={'attributes': {'value': 'user@example.com'}}, timestamp=1640995200850), DomMutation(mutation_type='text', target_id=9, details={'text': 'user@example.com'}, timestamp=1640995200850), DomMutation(mutation_type='attribute', target_id=10, details={'attributes': {'class': 'btn-clicked'}}, timestamp=1640995201300), DomMutation(mutation_type='add', target_id=12, details={'parent_id': 7, 'tag': 'div', 'attributes': {'class': 'success-message', 'data-testid': 'success-msg'}, 'text': ''}, timestamp=1640995201700)], 'interactions': [UserInteraction(action='click', target_id=9, value={'x': 140, 'y': 170}, timestamp=1640995200250), UserInteraction(action='input', target_id=9, value={'value': 'user@example.com', 'checked': False}, timestamp=1640995200800), UserInteraction(action='click', target_id=10, value={'x': 210, 'y': 230}, timestamp=1640995201100), UserInteraction(action='scroll', target_id=7, value={'x': 0, 'y': 100}, timestamp=1640995201500), UserInteraction(action='click', target_id=12, value={'x': 320, 'y': 360}, timestamp=1640995202100)], 'ui_nodes': {9: {'tag': 'input', 'aria_label': 'Email Address', 'data_testid': 'email-input', 'role': None, 'text': 'user@example.com', 'dom_path': 'unknown > html > body.main-content > form#contact-form > input#email'}, 10: {'tag': 'button', 'aria_label': 'Submit Form', 'data_testid': 'submit-btn', 'role': None, 'text': None, 'dom_path': 'unknown > html > body.main-content > form#contact-form > button#submit.btn-clicked'}, 12: {'tag': 'div', 'aria_label': None, 'data_testid': 'success-msg', 'role': None, 'text': None, 'dom_path': 'unknown > html > body.main-content > div.success-message'}, 7: {'tag': 'body', 'aria_label': None, 'data_testid': None, 'role': None, 'text': None, 'dom_path': 'unknown > html > body.main-content'}}, 'scroll_patterns': [ScrollPattern(scroll_event={'type': 3, 'data': {'source': 3, 'id': 7, 'x': 0, 'y': 100}, 'timestamp': 1640995201500}, mutation_event={'type': 3, 'data': {'source': 0, 'texts': [], 'attributes': [], 'removes': [], 'adds': [{'parentId': 7, 'nextId': None, 'node': {'type': 'Element', 'tagName': 'div', 'attributes': {'class': 'success-message', 'data-testid': 'success-msg'}, 'childNodes': [{'type': 'Text', 'textContent': 'Form submitted successfully!', 'id': 13}], 'id': 12}}]}, 'timestamp': 1640995201700}, delay_ms=200)]}, metadata={'num_events': 8, 'duration_ms': 1850, 'snapshot_before': {'type': 2, 'data': {'node': {'type': 'Document', 'id': 1, 'childNodes': [{'type': 'DocumentType', 'name': 'html', 'publicId': '', 'systemId': '', 'id': 2}, {'type': 'Element', 'tagName': 'html', 'attributes': {'lang': 'en'}, 'childNodes': [{'type': 'Element', 'tagName': 'head', 'attributes': {}, 'childNodes': [{'type': 'Element', 'tagName': 'title', 'attributes': {}, 'childNodes': [{'type': 'Text', 'textContent': 'Sample Page', 'id': 6}], 'id': 5}], 'id': 4}, {'type': 'Element', 'tagName': 'body', 'attributes': {'class': 'main-content'}, 'childNodes': [{'type': 'Element', 'tagName': 'form', 'attributes': {'id': 'contact-form'}, 'childNodes': [{'type': 'Element', 'tagName': 'input', 'attributes': {'type': 'email', 'id': 'email', 'data-testid': 'email-input', 'aria-label': 'Email Address', 'value': 'user@example.com'}, 'childNodes': [], 'id': 9}, {'type': 'Element', 'tagName': 'button', 'attributes': {'type': 'submit', 'id': 'submit', 'data-testid': 'submit-btn', 'aria-label': 'Submit Form', 'class': 'btn-clicked'}, 'childNodes': [{'type': 'Text', 'textContent': 'Submit', 'id': 11}], 'id': 10}], 'id': 8}], 'id': 7}], 'id': 3}]}, 'initialOffset': {'left': 0, 'top': 0}}, 'timestamp': 1640995200000}, 'session_id': 'sample-session'})
# ---
